CAN_in_the_middle/CAN_in_the_middle.ino

```diff
@@ -0,0 +1,278 @@
+        buf[3] = canId & 0xff;
+        // FIXME: grab extflags and put in here.
+        
+        send(buf, CMD_CAN_RECV, len + 4);
+        send(buf, CMD_ISO_RECV, len+4);
+    }
+
+
+    // handle Serial-incoming data    
+    for (count=Serial.available(); count>0; count--)
+    {
+        //logHexStr((INT32U)count, "readloopcount", 13);
+        inbuf[inbufidx++] = Serial.read();
+        if (inbufidx==1)
+        {
+            inbufcount = inbuf[0];
+            //logHexStr(inbufcount, "Setting inbufcount", 17);
+        }
+        
+        if (inbufidx == inbufcount)
+        // we've got enough of the packet
+        {
+            count = 1;  // exit the loop
+        }        
+    }
+        
+    // if we've received an entire message, process it here
+    if (inbufidx && inbufidx == inbufcount)
+    {
+        int fail_veh = -1;
+        int fail_iso = -1;
+        INT8U results_veh = CAN_FAIL;
+        INT8U results_iso = CAN_FAIL;
+
+        switch (inbuf[1])  // cmd byte
+        {
+            case CMD_CHANGE_BAUD:
+                Serial.begin(*(unsigned int*)(inbuf+2));
+                while (!Serial);
+                send(&results, CMD_CHANGE_BAUD_RESULT, 1);
+                break;
+                
+            case CMD_PING:
+                send(inbuf+2, CMD_PING_RESPONSE, inbufcount-2);
+                break;
+                
+            case CMD_CAN_BAUD:
+                baud = inbuf[2];
+KEEP_TRYING_ISO:
+                if(CAN_OK == ISO_CAN.begin(baud))                   // init can bus : baudrate = 500k
+                {
+                    Serial.write("@\x05\x01INIT", 7);
+                }
+                else
+                {
+                    Serial.write("@\x05\x02""FAIL", 7);
+                    delay(100);
+                    goto KEEP_TRYING_ISO;
+                }
+    
+KEEP_TRYING_VEH:
+                if(CAN_OK == VEH_CAN.begin(baud))                   // init can bus : baudrate = 500k
+                {
+                    Serial.write("@\x05\x01INIT", 7);
+                }
+                else
+                {
+                    Serial.write("@\x05\x02""FAIL", 7);
+                    delay(100);
+                    goto KEEP_TRYING_VEH;
+                }
+                while (!Serial);
+                break;
+                
+            case CMD_CAN_SEND:
+                //log("sending", 7);
+                // len, cmd, canid, canid2, extflag
+                canId = (INT32U)inbuf[5] | 
+                        ((INT32U)inbuf[4] << 8) |
+                        ((INT32U)inbuf[3] << 16) |
+                        ((INT32U)inbuf[2] << 24);
+                extflag = inbuf[6];
+                len = inbuf[0] - 7;
+                while(results_veh != CAN_OK && 
+                      results_iso != CAN_OK &&
+                      fail_veh < 10 &&
+                      fail_iso < 10)
+                {
+                    if(results_veh != CAN_OK)
```

Comment:
this is a little brutish.  we'll want to make these results/retrans independent of each other.